ocpbugs-8882: configure an additional clientca for the openshiftapi s… (#66833)
openshift/openshift-docs: modules/configure-an-additional-clientCA.adoc

Reviewer: This module looks like it is extra and should be deleted.

Author: Left in place, but redone to incorporate Jeana's suggestion below about using content for an intro.

// Module included in the following assemblies:
//
// * security/certificates/api-server.adoc

:_content-type: PROCEDURE
[id="configure-an-additional-clientCA-for-the-OpenShift-API-server_{context}"]

= Configure an additional clientCA for the OpenShift API server

Optionally, you may choose to invalidate the installer-generated kubeconfig. You would do this when:

> * You don't trust who installed the cluster

Reviewer 2: This list is wrongly formatted somehow, so it doesn't look well in the preview

Author: Seems to be OK now.

* The kubeconfig is leaked
* Other security-related needs exist, such as periodic rotation of the kubeconfig

To replace the installer-generated kubeconfig, remove the installer-generated clientCA from the API server

. Import an additional CA certificate in a configmap of the openshift-config namespace. The CA file must be in PEM format.
+
[source,terminal]
----
oc create configmap client-ca-custom -n openshift-config --from-file=ca-bundle.crt=ca.crt
----
+
. Patch the APIServer instance.
+
[source, terminal]
----
oc patch apiserver cluster --type=merge -p '{"spec": {"clientCA": {"name": "client-ca-custom"}}}'
----

. Test the new clientCA certificate with a certificate signed from the new clientCA.

> . If the test is successful, you can remove the installer-generated clientCA.

Reviewer 2: You are not explaining how to remove the installer-generated client CA yet you ask to remove it.

Reviewer 2: I'd remove this reference from this section but add a new section that explains the removal, after this section, and that suggests that you should first add the new CA as per the section above before replacing.

Pietro may provide a better view on how it should exactly look like

Author: Thanks for the info, Pablo. Could you or Pietro send me the details on how to remove the installer-generated client CA?

Reviewer 3: @nalhadef here is a bit confusing, this KCS explains the procedure: https://access.redhat.com/solutions/6054981

To give you more context:
- the installer-generate kubeconfig contains a client TLS certificate signed by a CA
- the CA signing the kubeconfig is stored into the configmap named `admin-kubeconfig-client-ca` into the `openshift-config` namespace
- in order to invalidate the installer-generated kubeconfig you have to replace the CA (thus the content of the configmap)
- replacing the CA can be dangerous because if you are using the installer-generated kubeconfig and the content of the CM is somehow invalid you can be locked out from the API
- for that reason in the KCS I suggest to:
   - first add an additional CA
   - generate a new kubeconfig using the new CA
   - replace the original CA replacing the configmap.

Author: Incorporated KCS article content

// Module included in the following assemblies:
//
// * security/certificates/api-server.adoc

:_content-type: PROCEDURE
[id="configure-an-additional-clientCA-for-the-OpenShift-API-server_{context}"]

> = Configure an additional clientCA for the OpenShift API server

Reviewer 4: This heading doesn't seem to fit. Aren't we replacing/invalidating the kubeconfig? Isn't that the point of the procedure?

Author: We're invalidating the installer-generated kubeconfig and replacing it with a newly generated obe. The new one must be added and configured before the installer-generated one is invalidated. (The existing one can't be removed, it can be invalidated, though.)

> Optionally, you may choose to invalidate the installer-generated kubeconfig. You would do this when:

Reviewer 4: ```suggestion
Optionally, you can invalidate the installer-generated kubeconfig. You might do this if any of the following conditions exist:
```

Reviewer 4: Is _invalidate_ the right word? I should think _replace_, which you use later is better. If you invalidate the kubeconfig, there is no kubeconfig, right?

Author: used suggested text, including "replace."

> * You don't trust who installed the cluster

Reviewer 4: ```suggestion
* You do not trust who installed the cluster
```

Reviewer 4: Why would you not trust who installed the cluster?

Author: - Spelled out "do not."
- If you don't know the certifying authority, you might feel more secure generating a certificate from a known and trusted CA.

> * The kubeconfig is leaked

Reviewer 4: ```suggestion
* The kubeconfig is leaked.
```

Reviewer 4: Not sure what this means, maybe just my ignorance.

Author: - Added periods at the end of the listed items.
- I don't know how a kubeconfig could be leaked, either.

> * Other security-related needs exist, such as periodic rotation of the kubeconfig

Reviewer 4: ```suggestion
* Other security-related needs exist, such as the periodic rotation of the kubeconfig.
```

Author: Made suggested change

> 

Reviewer 4: .Procedure

Author: Added

> To replace the installer-generated kubeconfig, remove the installer-generated clientCA from the API server

Reviewer 4: ```suggestion
To replace the installer-generated kubeconfig, remove the installer-generated clientCA from the API server:
```

Author: done

> . Import an additional CA certificate in a configmap of the openshift-config namespace. The CA file must be in PEM format.

Reviewer 4: ```suggestion
. Import an additional CA certificate in a configmap in the openshift-config namespace. The CA file must be in PEM format.
```

Reviewer 4: ```suggestion
. Import an additional CA certificate in a configmap of the openshift-config namespace. The CA file must be in PEM format.
```

Author: Changed

+
[source,terminal]
----
oc create configmap client-ca-custom -n openshift-config --from-file=ca-bundle.crt=ca.crt
----
+

> . Patch the APIServer instance.

Reviewer 4: ```suggestion
. Use the following command to patch the APIServer instance:
```

Author: changed.
